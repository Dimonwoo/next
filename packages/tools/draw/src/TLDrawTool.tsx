import { IdleState, PointingState, CreatingState } from './states'
import { TLShapeProps, TLTool, TLApp, TLShape } from '@tldraw/core'
import type { TLDrawShapeProps, TLDrawShape } from '@tldraw/draw-shape'

export abstract class TLDrawTool<
  T extends TLDrawShape = TLDrawShape,
  S extends TLShape = TLShape,
  R extends TLApp<S> = TLApp<S>
> extends TLTool<S, R> {
  static id = 'draw'

  static states = [IdleState, PointingState, CreatingState]

  static initial = 'idle'

  /** Whether to simplify the shape's points after creating. */
  simplify = true

  /** The minimum distance between points when simplifying a line. */
  simplifyTolerance = 1

  abstract shapeClass: {
    new (props: TLShapeProps & Partial<TLDrawShapeProps & unknown>): T
  }
}
