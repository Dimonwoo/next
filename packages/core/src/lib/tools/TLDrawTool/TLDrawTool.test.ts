import { TLDrawTool } from './TLDrawTool'
import { TLApp, TLDrawShape } from '~lib'

describe('A minimal test', () => {
  it('Creates the shape', () => {
    class DrawTool extends TLDrawTool<TLDrawShape> {
      static id = 'draw'
      static shortcut = ['d']
      Shape = TLDrawShape
    }
    const app = new TLApp()
    const shape = new DrawTool(app, app)
    expect(shape).toBeDefined()
  })
})
