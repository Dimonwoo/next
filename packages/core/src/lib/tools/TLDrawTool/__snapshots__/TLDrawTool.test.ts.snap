// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When using the tool Creates a shape and transitions to select.idle after pointer up: created line 1`] = `
TLLineShape {
  "clone": [Function],
  "getBounds": [Function],
  "getCachedSerialized": [Function],
  "getCenter": [Function],
  "getRotatedBounds": [Function],
  "getSerialized": [Function],
  "hideContextBar": false,
  "hideResizeHandles": false,
  "hideRotateHandle": false,
  "hideSelection": false,
  "hideSelectionDetail": false,
  "hitTestBounds": [Function],
  "hitTestLineSegment": [Function],
  "hitTestPoint": [Function],
  "isAspectRatioLocked": false,
  "isDirty": true,
  "isEditable": false,
  "isStateful": false,
  "lastSerialized": Object {},
  "nonce": 0,
  "normalizedHandles": Array [],
  "onHandleChange": [Function],
  "onResize": [Function],
  "onResizeStart": [Function],
  "props": Object {
    "handles": Array [
      Object {
        "point": Array [
          0,
          0,
        ],
      },
      Object {
        "point": Array [
          1,
          1,
        ],
      },
    ],
    "id": "test_line",
    "parentId": "page1",
    "point": Array [
      100,
      100,
      0.5,
    ],
    "type": "line",
  },
  "showCloneHandles": false,
  "stayMounted": false,
  "type": "line",
  "validateProps": [Function],
  "version": 1,
}
`;

exports[`When using the tool Transitions to creating and creates a dot shape on pointer down: created dot 1`] = `
TLDotShape {
  "clone": [Function],
  "getBounds": [Function],
  "getCachedSerialized": [Function],
  "getCenter": [Function],
  "getRotatedBounds": [Function],
  "getSerialized": [Function],
  "hideContextBar": false,
  "hideResizeHandles": true,
  "hideRotateHandle": true,
  "hideSelection": true,
  "hideSelectionDetail": true,
  "hitTestBounds": [Function],
  "hitTestLineSegment": [Function],
  "hitTestPoint": [Function],
  "isAspectRatioLocked": false,
  "isDirty": true,
  "isEditable": false,
  "isStateful": false,
  "lastSerialized": Object {},
  "nonce": 0,
  "onHandleChange": [Function],
  "onResize": [Function],
  "props": Object {
    "id": "test_dot",
    "parentId": "page1",
    "point": Array [
      94,
      94,
    ],
    "radius": 6,
    "type": "dot",
  },
  "showCloneHandles": false,
  "stayMounted": false,
  "type": "dot",
  "validateProps": [Function],
  "version": 1,
}
`;
